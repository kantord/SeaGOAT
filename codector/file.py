import time


class File:
    def __init__(self, path: str):
        self.path = path
        self.score = 0.0
        self.commit_messages = []

    def _increment_score(self, committed_date):
        current_time = int(time.time())
        age_of_commit_in_seconds = current_time - committed_date
        age_of_commit_in_days = int(age_of_commit_in_seconds / 86400)
        self.score += 100 / (age_of_commit_in_days**2)

    def _add_commit_message(self, message: str):
        if message not in self.commit_messages:
            self.commit_messages.append(message)

    def add_commit(self, commit):
        self._add_commit_message(commit.message)
        self._increment_score(commit.committed_date)

    def get_metadata(self):
        commit_messages = "\n".join(self.commit_messages)
        return f"""###
{commit_messages}"""
