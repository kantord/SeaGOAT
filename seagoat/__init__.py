__version__ = "0.49.4"
