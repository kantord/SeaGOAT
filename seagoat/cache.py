import hashlib
import os
import pickle
from pathlib import Path
from typing import Generic
from typing import TypeVar

import appdirs

T = TypeVar("T")

CACHE_FORMAT_VERSION = 15


class Cache(Generic[T]):
    def __init__(self, cache_name: str, path: Path, initial_value: T):
        self._path = path
        self.data = initial_value
        self._cache_name = cache_name

    def load(self):
        try:
            with open(self._get_cache_file(), "rb") as cache_file:
                self.data = pickle.load(cache_file)
        except (FileNotFoundError, pickle.UnpicklingError, EOFError):
            pass

    def persist(self):
        with open(self._get_cache_file(), "wb") as cache_file:
            pickle.dump(self.data, cache_file)

    def _get_cache_file(self):
        return self.get_cache_folder() / self._cache_name

    def get_cache_folder(self):
        cache_folder = self._get_cache_root() / self._get_project_hash()
        cache_folder.mkdir(parents=True, exist_ok=True)

        return cache_folder

    def _get_cache_root(self):
        if "RUNNER_TEMP" in os.environ:
            return Path(os.environ["RUNNER_TEMP"])

        return Path(
            appdirs.user_cache_dir(
                "seagoat-pytest" if "PYTEST_CURRENT_TEST" in os.environ else "seagoat"
            )
        )

    def _get_project_hash(self):
        normalized_path = Path(self._path).expanduser().resolve()
        text = f"""
        Cache version: {CACHE_FORMAT_VERSION}
        Normalized path: {normalized_path}
        """

        return hashlib.sha256(text.encode()).hexdigest()
