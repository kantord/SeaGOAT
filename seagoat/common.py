IGNORED_BRANCHES = {"gh-pages"}
SUPPORTED_FILE_TYPES = {
    ".txt",
    ".md",
    ".py",
    ".c",
    ".cpp",
    ".h",
    ".hpp",
    ".ts",
    ".js",
    ".tsx",
    ".jsx",
    ".html",
}
