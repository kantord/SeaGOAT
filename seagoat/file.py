import hashlib
import time
from typing import List
from typing import Literal

from chromadb.errors import Dict


class File:
    def __init__(self, path: str, absolute_path: str):
        self.path = path
        self.absolute_path = absolute_path
        self._commit_times = set()
        self.commit_messages = set()

    def __repr__(self):
        return f"<File {self.path} {self.get_score()}>"

    def get_score(self):
        current_time = int(time.time())

        return sum(
            1000 / max(int((current_time - committed_date) / 86400) ** 2, 1)
            for committed_date in self._commit_times
        )

    def _add_commit_message(self, message: str):
        self.commit_messages.add(message)

    def add_commit(self, commit):
        self._commit_times.add(commit.committed_date)
        self._add_commit_message(commit.message)

    def get_metadata(self):
        commit_messages = "\n-".join(sorted(self.commit_messages))
        return f"""###
    Filename: {self.path}
    Commits:
{commit_messages}"""

    def _get_file_lines(self) -> Dict[int, str]:
        with open(self.absolute_path, "r", encoding="utf-8") as source_code_file:
            lines = {
                (i + 1): line
                for i, line in enumerate(source_code_file.read().splitlines())
            }

        return lines

    def _format_chunk_summary(self, relevant_lines: List[str]):
        truncated_lines = [line[:100] for line in relevant_lines]
        chunk = "\n".join(truncated_lines)
        chunk = chunk + self.get_metadata()

        return chunk

    def _get_context_lines(
        self, lines: Dict[int, str], line_number: int, direction: Literal[-1, 1]
    ) -> List[str]:
        context_lines = []
        for i in range(1, 6):
            current_line_number = line_number + (direction * i)
            current_line = lines.get(current_line_number)

            if current_line is None:
                break

            if direction == -1:
                context_lines = [current_line] + context_lines
            else:
                context_lines.append(current_line)

            if self._line_has_relevant_data(current_line):
                break

        return context_lines

    def _get_chunk_for_line(self, line_number: int, lines: Dict[int, str]):
        relevant_lines = (
            self._get_context_lines(lines, line_number, -1)
            + [lines[line_number]]
            + self._get_context_lines(lines, line_number, 1)
        )
        return FileChunk(self, line_number, self._format_chunk_summary(relevant_lines))

    def _line_has_relevant_data(self, line: str):
        return sum(c.isalnum() for c in line) > 3

    def get_chunks(self):
        try:
            lines = self._get_file_lines()
            return [
                self._get_chunk_for_line(line_number, lines)
                for line_number in lines.keys()
                if self._line_has_relevant_data(lines[line_number])
            ]

        except FileNotFoundError:
            return []


# pylint: disable=too-few-public-methods
class FileChunk:
    def __init__(self, parent: File, codeline: int, chunk: str):
        self.path = parent.path
        self.codeline = codeline
        self.chunk = chunk
        self.chunk_id = self._get_id()

    def _get_id(self):
        text = f"""
        Path: {self.path}
        Code line: {self.codeline}
        Chunk: {self.chunk}
        """
        return hashlib.sha256(text.encode()).hexdigest()
