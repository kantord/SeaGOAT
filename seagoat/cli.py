import os
from functools import cache

import click
import requests
from pygments import highlight
from pygments.formatters import TerminalFormatter
from pygments.lexers import get_lexer_for_filename
from pygments.lexers.javascript import JavascriptLexer
from pygments.lexers.javascript import TypeScriptLexer

from seagoat.server import get_server_info_file
from seagoat.server import load_server_info


def query_server(query, server_address):
    response = requests.get(f"{server_address}/query/{query}")
    try:
        response.raise_for_status()
    except requests.exceptions.HTTPError as err:
        print("HTTP error occurred:")
        print(f"Response code: {response.status_code}")
        print(f"Error: {err}")
        print(f"Response body: {response.text}")
        raise
    return response.json()["results"]


@cache
def get_highlighted_lines(file_name: str):
    with open(file_name, "r", encoding="utf-8") as source_code_file:
        code = source_code_file.read()

    if file_name.endswith(".md"):
        return code.splitlines()

    if file_name.endswith(".jsx"):
        lexer = JavascriptLexer()
    elif file_name.endswith(".tsx"):
        lexer = TypeScriptLexer()
    else:
        lexer = get_lexer_for_filename(file_name)

    result = highlight(code, lexer, TerminalFormatter())

    return result.splitlines()


def print_result_line(result, line, color_enabled):
    if color_enabled:
        highlighted_lines = get_highlighted_lines(str(result["full_path"]))
        print(
            f"{result['path']}:{click.style(str(line), bold=True)}:{highlighted_lines[line - 1]}"
        )
    else:
        for line_content in result["lines"]:
            if line_content["line"] == line:
                print(f"{result['path']}:{line}:{line_content['line_text']}")
                break


@click.command()
@click.argument("query")
@click.argument("repo_path", required=False, default=os.getcwd())
@click.option(
    "--no-color",
    is_flag=True,
    help="Disable formatting. Automatically enabled when part of a bash pipeline.",
)
def seagoat(query, repo_path, no_color):
    """
    Query your codebase for your QUERY in the Git repository REPO_PATH.

    Your query can either be text that you expect to find in a file,
    or a description of what you are looking for.

    When REPO_PATH is not specified, the current working directory is
    assumed to be the repository path.
    """
    _, __, ___, server_address = load_server_info(get_server_info_file(repo_path))
    results = query_server(query, server_address)

    color_enabled = os.isatty(0) and not no_color
    for result in results:
        for result_line in result.get("lines", []):
            print_result_line(result, result_line["line"], color_enabled)


if __name__ == "__main__":
    # pylint: disable-next=no-value-for-parameter
    seagoat()
