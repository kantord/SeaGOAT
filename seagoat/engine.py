"""
    This module allows you to use seagoat as a library
"""
import asyncio
from concurrent.futures import ThreadPoolExecutor
from functools import partial
from itertools import chain
from pathlib import Path
from typing import Dict
from typing import List
from typing import Set

import nest_asyncio
from tqdm import tqdm
from typing_extensions import TypedDict

from seagoat.cache import Cache
from seagoat.file import File
from seagoat.repository import Repository
from seagoat.sources import chroma
from seagoat.sources import ripgrep


RepositoryData = TypedDict(
    "RepositoryData",
    {
        "last_analyzed_version_of_branch": Dict[str, str],
        "required_commits": Set[str],
        "commits_already_analyzed": Set[str],
        "file_data": Dict[str, File],
        "sorted_files": List[str],
        "chunks_already_analyzed": Set[str],
    },
)


nest_asyncio.apply()


class Engine:
    """
    A search engine for a code repository
    """

    def __init__(self, path: str):
        """
        Initializes the library
        """
        self.path = path
        self.query_string = ""
        self._results_from_chromadb = []
        self._results = []
        self._cache = Cache[RepositoryData](
            "cache",
            Path(path),
            {
                "last_analyzed_version_of_branch": {},
                "required_commits": set(),
                "commits_already_analyzed": set(),
                "file_data": {},
                "sorted_files": [],
                "chunks_already_analyzed": set(),
            },
        )
        self._cache.load()
        self.repository = Repository(path, self._cache)
        self._fetchers = {
            "async": [
                ripgrep.initialize(self.repository),
            ],
            "sync": [
                chroma.initialize(self.repository),
            ],
        }

    def analyze_codebase(self):
        self.repository.analyze_files()
        self._create_vector_embeddings()

    def _add_to_collection(self, chunk):
        for source in chain(*self._fetchers.values()):
            source["cache_chunk"](chunk)

    def _create_vector_embeddings(self):
        chunks_to_process = []
        minimum_files_to_analyze = max(40, int(len(self.repository.top_files()) * 0.2))
        for file in self.repository.top_files()[:minimum_files_to_analyze]:
            for chunk in file.get_chunks():
                chunks_to_process.append(chunk)
        for chunk in tqdm(chunks_to_process, desc="Analyzing source code"):
            if chunk.chunk_id in self._cache.data["chunks_already_analyzed"]:
                continue

            self._add_to_collection(chunk)
            self._cache.data["chunks_already_analyzed"].add(chunk.chunk_id)

        self._cache.persist()

        for source in chain(*self._fetchers.values()):
            source["persist"]()

    def query(self, query: str):
        self.query_string = query

    async def fetch(self):
        self._results = []
        executor = ThreadPoolExecutor(max_workers=1)
        loop = asyncio.get_event_loop()
        async_tasks = [
            loop.run_in_executor(executor, partial(source["fetch"], self.query_string))
            for source in self._fetchers["async"]
        ]

        for source in self._fetchers["sync"]:
            self._results.extend(source["fetch"](self.query_string))

        results = await asyncio.gather(*async_tasks)
        for result in results:
            self._results.extend(result)

    def fetch_sync(self):
        loop = asyncio.get_event_loop()
        loop.run_until_complete(self.fetch())

    def get_results(self):
        merged_results = {}

        for result_item in self._results:
            if result_item.path not in merged_results:
                merged_results[result_item.path] = result_item
                continue

            merged_results[result_item.path].extend(result_item)

        sorted_by_score = sorted(
            merged_results.values(),
            key=lambda x: x.get_best_score(self.query_string),
        )
        position_by_result = {item: i for i, item in enumerate(sorted_by_score)}
        position_by_file = {
            file.path: i for i, file in enumerate(self.repository.top_files())
        }

        return list(
            sorted(
                merged_results.values(),
                key=lambda x: (
                    0.8 * position_by_result[x] + 0.2 * position_by_file[x.path]
                ),
            )
        )
