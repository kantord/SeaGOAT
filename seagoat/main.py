from functools import cache
import os
import sys
import click
from pygments import highlight
from pygments.lexers import get_lexer_for_filename
from pygments.formatters import TerminalFormatter
from pygments.lexers.javascript import JavascriptLexer, TypeScriptLexer

from seagoat.engine import Engine


@cache
def get_highlighted_lines(file_name: str):
    with open(file_name, "r", encoding="utf-8") as source_code_file:
        code = source_code_file.read()

    if file_name.endswith(".md"):
        return code.splitlines()

    if file_name.endswith(".jsx"):
        lexer = JavascriptLexer()
    elif file_name.endswith(".tsx"):
        lexer = TypeScriptLexer()
    else:
        lexer = get_lexer_for_filename(file_name)

    result = highlight(code, lexer, TerminalFormatter())

    return result.splitlines()


def print_result_line(result, line, color_enabled):
    if color_enabled:
        highlighted_lines = get_highlighted_lines(str(result.full_path))
        print(
            f"{result.path}:{click.style(str(line), bold=True)}:{highlighted_lines[line - 1]}"
        )
    else:
        print(f"{result.path}:{line}:{result.line_texts[line - 1]}")


@click.command()
@click.argument("query")
@click.argument("repo_path", required=False, default=os.getcwd())
@click.option("--no-color", is_flag=True)
def seagoat(query, repo_path, no_color):
    """Query your codebase using vector embeddings"""
    my_seagoat = Engine(repo_path)
    my_seagoat.analyze_codebase()

    my_seagoat.query(query)
    my_seagoat.fetch_sync()
    results = my_seagoat.get_results()

    color_enabled = sys.stdout.isatty() and not no_color
    for result in results:
        for line in result.get_lines(query):
            print_result_line(result, line, color_enabled)


if __name__ == "__main__":
    # pylint: disable-next=no-value-for-parameter
    seagoat()
