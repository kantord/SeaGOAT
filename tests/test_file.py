# pylint: disable=redefined-outer-name
from pathlib import Path

import pytest

from seagoat.file import File


@pytest.fixture
def repo_folder(repo):
    with open(Path(repo.working_dir) / "hello.md", "w", encoding="utf-8") as hello_file:
        hello_file.write("Hello world!")
    yield Path(repo.working_dir)


def test_file_returns_global_metadata_1(repo_folder, snapshot):
    my_file = File(
        "hello.md",
        repo_folder / "hello.md",
        0.543245,
        [
            "First commit",
            "Second commit",
        ],
    )

    assert my_file.get_metadata() == snapshot


def test_file_returns_global_metadata_2(repo_folder, snapshot):
    my_file = File("hello.md", repo_folder / "hello.md", 0.234234, ["unrelated commit"])

    assert my_file.get_metadata() == snapshot


def test_ignores_almost_empyt_lines_in_chunks(repo):
    repo.add_file_change_commit(
        file_name="example.py",
        contents="""#this is a Python file

# xd

class FooBar:


def __init__(self):
        pass""",
        author=repo.actors["John Doe"],
        commit_message=".",
    )

    my_file = File(
        "example.py", str(Path(repo.working_dir) / "example.py"), 0.234234, []
    )
    assert {item.codeline for item in my_file.get_chunks()} == {1, 5, 8, 9}
    line5 = [item for item in my_file.get_chunks() if item.codeline == 5][0]
    found_lines = line5.chunk.split("###")[0].splitlines()
    assert found_lines == [
        "#this is a Python file",
        "",
        "# xd",
        "",
        "class FooBar:",
        "",
        "",
        "def __init__(self):",
    ]
