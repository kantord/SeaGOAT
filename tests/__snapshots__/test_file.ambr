# serializer version: 1
# name: test_file_returns_global_metadata_1
  '''
  ###
      Filename: hello.md
      Commits:
  First commit
  -Second commit
  '''
# ---
# name: test_file_returns_global_metadata_2
  '''
  ###
      Filename: hello.md
      Commits:
  unrelated commit
  '''
# ---
